### PR TITLE
Implement stateless mode for Thanos Ruler (continue #4250)

### PR DESCRIPTION
**Pull request by yeya24**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

This pr is based on top of the awesome pr https://github.com/thanos-io/thanos/pull/4250 by @idoqo and fixes E2E tests.

## Verification

E2E tests should pass.

### CODE REVIEW

**Comment by yeya24** [comment]
https://github.com/prometheus/prometheus/pull/9109 requires some modification on the vendored grafana agent wal package. 

```
/home/runner/work/thanos/thanos/pkg/rules/remotewrite/wal_test.go:75:29: cannot use &collector (value of type *walDataCollector) as wal.WriteTo value in struct literal: missing method UpdateSeriesSegment
```

Maybe let's wait for the upgrade from the agent side? @rfratto

**Comment by rfratto** [comment]
> Maybe let's wait for the upgrade from the agent side? @rfratto

I'm going to be bumping the Prometheus dependency in grafana/agent ideally this week and expect to also run into this. I'll tag you @yeya24 once we have the PR to update the code to resolve the error.

**Comment by rfratto** [comment]
> I'm going to be bumping the Prometheus dependency in grafana/agent ideally this week and expect to also run into this. I'll tag you @yeya24 once we have the PR to update the code to resolve the error.

Hey y'all, I wanted to give a quick update on this. I've put effort into updating the Grafana Agent's Prometheus dependency this week, but it's going to take some more time. The bump also requires a bump to github.com/prometheus/common, which is currently blocked by a few of the exporters we embed still using the (since removed) github.com/prometheus/common/log package. Such is the life of a project that has a lot of dependencies :) 

We've definitely made progress here, but it might take another week before I can update the WAL code. I'll still tag y'all when we have something.

File: CHANGELOG.md
Comment (GiedriusS) [inline CHANGELOG.md:21]: Maybe we could have a more descriptive CHANGELOG message such as: "added stateless mode to ruler according to https://github.com/thanos-io/thanos/blob/main/docs/proposals/202005_scalable-rule-storage.md#proposal"?

File: pkg/rules/remotewrite/wal_test.go
Comment (GiedriusS) [inline pkg/rules/remotewrite/wal_test.go:21]: Let's convert to `testutil.Ok()` and so on, and remove `testify` as we have those helper functions already in Thanos

File: pkg/rules/remotewrite/wal_test.go
Comment (GiedriusS) [inline pkg/rules/remotewrite/wal_test.go:43]:
```suggestion
func TestStorage_TruncateAfterClose(t *testing.T) {
```

File: test/e2e/rule_test.go
Comment (GiedriusS) [inline test/e2e/rule_test.go:530]: Do we need to implement this before merging?

**Review by GiedriusS — COMMENTED** [review]
Just some nits and questions. Good work! Do we still need `pkg/rules/remotewrite/{wal,series}.go` with https://github.com/prometheus/prometheus/pull/8785 merged? ;P It says:
```
// TODO(idoqo): Migrate to prometheus package when https://github.com/prometheus/prometheus/pull/8785 is ready.
```

**Comment by yeya24** [comment]
> Just some nits and questions. Good work! Do we still need `pkg/rules/remotewrite/{wal,series}.go` with [prometheus/prometheus#8785](https://github.com/prometheus/prometheus/pull/8785) merged? ;P It says:

It seems that we can totally remove the `pkg/rules/remotewrite` package after using the upstream Prometheus agent code 😄

File: test/e2e/rule_test.go
Comment (yeya24) [reply test/e2e/rule_test.go:530]: I guess this is not a blocker as we are using the upstream prom agent code already so data should be persisted. Not sure how to implement this right now though.

File: CHANGELOG.md
Comment (yeya24) [reply CHANGELOG.md]: Done

**Comment by yeya24** [comment]
@GiedriusS Let me know if there is anything left in this pr. If it is small nits then I guess we can clean up later. 😄

File: cmd/thanos/rule.go
Comment (GiedriusS) [inline cmd/thanos/rule.go:364]: Could we please call this something other than `db` to avoid shadowing the outer variable? This is confusing and I had to double check this place :joy: I suggest `agentDB` or something like that, and to rename the outer `db` variable into `tsdbDB`.

File: cmd/thanos/rule.go
Comment (GiedriusS) [inline cmd/thanos/rule.go:589]: In the future maybe we could also answer StoreAPI queries by using the data in the WAL? Perhaps worth adding a TODO here :smile:

File: cmd/thanos/rule.go
Comment (GiedriusS) [inline cmd/thanos/rule.go:592]: I think this is an outdated comment that we can remove since we have the Rules server?

File: test/e2e/rule_test.go
Comment (GiedriusS) [inline test/e2e/rule_test.go:512]: We can just use `time.Now` directly without another function

File: test/e2e/rule_test.go
Comment (GiedriusS) [inline test/e2e/rule_test.go:508]: Here we should probably add some kind of `WaitSumMetricsWithOptions` to avoid random errors in tests so that it would wait until at least one remote write has happened ;P

File: test/e2e/rule_test.go
Comment (GiedriusS) [reply test/e2e/rule_test.go:530]: :+1:

**Review by GiedriusS — APPROVED** [review]
Just a few very small nits but after that, we can merge this IMO :+1:

File: cmd/thanos/rule.go
Comment (yeya24) [reply cmd/thanos/rule.go:589]: I am not sure about this. If the agent is able to query head data, then what's the difference between the agent and a prometheus with very low retention? I am not sure whether this is in their roadmap.

File: cmd/thanos/rule.go
Comment (yeya24) [reply cmd/thanos/rule.go:592]: Done

**Comment by yeya24** [comment]
Hi @GiedriusS, I have addressed the comments in the latest commit.

**Review by GiedriusS — APPROVED** [review]
In the end, it became a very small pull request with all of the agent code in upstream Prometheus :muscle: awesome!

File: cmd/thanos/rule.go
Comment (GiedriusS) [reply cmd/thanos/rule.go]: Fair point :+1:

**Comment by onprem** [comment]
Great work! :tada:

**Review by bwplotka — COMMENTED** [review]
Amazing work @yeya24 !

**Pull request merged**
